File: src/links/links_db_service.rs
use super::dto::Link;
use crate::utils::get_now_time;
use log::error;
use rusqlite::{params, Connection, Result};
use std::env;

pub struct LinksDbService {
    db_name: String,
}

impl LinksDbService {
    pub fn new() -> Self {
        let db_name = env::var("DB_NAME").expect("DB_NAME must be set");
        Self { db_name }
    }

    fn open_connection(&self) -> Result<Connection> {
        Connection::open(&self.db_name)
    }

    pub fn create_one(&self, path: &str, name: &str) -> Result<&str> {
        let conn = self.open_connection()?;

        return match conn.execute(
            "INSERT INTO links (path, name, is_reachable) VALUES (?, ?, 1)",
            params![path, name],
        ) {
            Ok(changes) => {
                if changes == 1 {
                    Ok("One path created")
                } else {
                    Ok("No path created")
                }
            }
            Err(e) => {
                error!("Error creating path: {}", e);
                Err(e)
            }
        };
    }

    pub fn get_all(&self, is_reachable: bool, show_duplicate: bool) -> Result<Vec<Link>> {
        let conn = self.open_connection()?;
        let query = if show_duplicate {
            "SELECT * FROM links WHERE is_reachable = ? AND duplicate_id IS NOT NULL ORDER BY is_downloaded"
        } else {
            "SELECT * FROM links WHERE is_reachable = ? AND duplicate_id IS NULL ORDER BY is_downloaded"
        };
        let mut stmt = conn.prepare(query)?;

        let rows = stmt.query_map([is_reachable], |row| {
            Ok(Link {
                id: row.get(0)?,
                path: row.get(1)?,
                name: row.get(2)?,
                is_downloaded: row.get(3)?,
                progress: row.get(4)?,
                downloaded_mediafiles: row.get(5)?,
                mediafiles: row.get(6)?,
                date_update: row.get(7)?,
                date_create: row.get(8)?,
                is_reachable: row.get(9)?,
                duplicate_id: row.get(10)?,
                duplicate_path: None,
            })
        })?;

        let result: Result<Vec<_>, _> = rows.collect();
        result
    }

    pub fn remove(&self, id: usize) -> Result<&str> {
        let conn = self.open_connection()?;

        return match conn.execute("DELETE FROM links WHERE id = ?", [id]) {
            Ok(changes) => {
                if changes == 1 {
                    Ok("One path removed")
                } else {
                    Ok("No path removed")
                }
            }
            Err(e) => {
                error!("Error removing path: {}", e);
                Err(e)
            }
        };
    }

    pub fn get_one(&self, id: usize) -> Result<Option<Link>> {
        let conn = self.open_connection()?;
        let mut stmt = conn.prepare("SELECT * FROM links WHERE id = ?")?;
        let mut rows = stmt.query([id])?;

        if let Some(row) = rows.next()? {
            let link = Link {
                id: row.get(0)?,
                path: row.get(1)?,
                name: row.get(2)?,
                is_downloaded: row.get(3)?,
                progress: row.get(4)?,
                downloaded_mediafiles: row.get(5)?,
                mediafiles: row.get(6)?,
                date_update: row.get(7)?,
                date_create: row.get(8)?,
                is_reachable: row.get(9)?,
                duplicate_id: row.get(10)?,
                duplicate_path: None,
            };
            Ok(Some(link))
        } else {
            Ok(None)
        }
    }

    pub fn tag_unreachable(&self, id: usize, is_reachable: bool) -> Result<String> {
        let conn = self.open_connection()?;
        let changes = conn.execute(
            "UPDATE links SET is_reachable = ?, date_update = ? WHERE id = ?",
            params![is_reachable, get_now_time(), id],
        )?;

        Ok(if changes == 1 {
            format!(
                "One path tagged as {}",
                if is_reachable {
                    "reachable"
                } else {
                    "unreachable"
                }
            )
        } else {
            "No path tagged".to_string()
        })
    }

    pub fn add_duplicate(&self, link_id: usize, duplicate_id: usize) -> Result<String> {
        let conn = self.open_connection()?;
        let changes = conn.execute(
            "UPDATE links SET duplicate_id = ?, date_update = ? WHERE id = ?",
            params![duplicate_id, get_now_time(), link_id],
        )?;

        Ok(if changes == 1 {
            format!(
                "Link id {} tagged as duplicate of {}",
                link_id, duplicate_id
            )
        } else {
            "No path tagged".to_string()
        })
    }

    pub fn update_files_number(
        &self,
        id: usize,
        mediafiles: usize,
        downloaded_mediafiles: usize,
        is_downloaded: bool,
        progress: usize,
    ) -> Result<String> {
        let conn = self.open_connection()?;

        let changes = conn.execute(
            "UPDATE links SET mediafiles = ?, downloaded_mediafiles = ?, is_downloaded = ?, progress = ?, date_update = ? WHERE id = ?",
            params![mediafiles, downloaded_mediafiles, is_downloaded, progress, get_now_time(), id],
        )?;
        Ok(if changes == 1 {
            "One path updated".to_string()
        } else {
            "No path updated".to_string()
        })
    }
}

File: backend_node/src/links/links.db.service.ts
import { HttpException, HttpStatus, Injectable } from '@nestjs/common';
import { AsyncDatabase } from 'promised-sqlite3';
import { iResult } from '../entities/common';
import { CreateLinkDto, Link, LinkDto, UpdateLinkDto } from '../entities/link.entity';
import { dateConvert, resultMaker } from '../helpers/common';

@Injectable()
export default class LinksDbService {
  public async getAll(isReachable: boolean, showDuplicate: boolean): Promise<Link[]> {
    const db = await AsyncDatabase.open(process.env.DB_NAME);
    let query: string;

    if (showDuplicate) {
      query = `
          SELECT l.*, d.path AS duplicate_path 
          FROM links AS l
          LEFT JOIN links AS d
          ON l.duplicate_id = d.id
          WHERE l.is_reachable = ? AND l.duplicate_id IS NOT NULL 
          ORDER BY l.is_downloaded
        `;
    } else {
      query = `
        SELECT * 
        FROM links 
        WHERE is_reachable = ? AND duplicate_id IS NULL 
        ORDER BY is_downloaded
        `;
    }

    const result = await db.all<Link>(query, [isReachable]);

    await db.close();

    const mappedResult = result.map((row) => new LinkDto(row));
    return mappedResult;
  }

  public async createOne({ path, name }: CreateLinkDto): Promise<iResult> {
    const db = await AsyncDatabase.open(process.env.DB_NAME);
    try {
      const result = await db.run(
        `
        INSERT INTO links (path, name, is_reachable)
        VALUES (?, ?, 1)`,
        [path.trim(), name],
      );

      await db.close();
      if (result.changes === 1) {
        return resultMaker('One path created');
      }
      return resultMaker('No path created');
    } catch (error) {
      if (error.code === 'SQLITE_CONSTRAINT') {
        throw new HttpException('Already exist ', HttpStatus.BAD_REQUEST);
      }
      throw new HttpException('can not make transaction', HttpStatus.INTERNAL_SERVER_ERROR);
    }
  }

  public async remove(id: string): Promise<iResult> {
    const db = await AsyncDatabase.open(process.env.DB_NAME);
    try {
      const result = await db.run(
        `
        DELETE FROM links
        WHERE id=?`,
        [id],
      );
      await db.close();
      if (result.changes === 1) {
        return resultMaker('One path removed');
      }
      return resultMaker('No one path removed');
    } catch (error) {
      throw new HttpException('can not make transaction', HttpStatus.INTERNAL_SERVER_ERROR);
    }
  }

  public async getOne(id: string): Promise<Link> {
    const db = await AsyncDatabase.open(process.env.DB_NAME);
    try {
      const result = await db.get<Link>(
        `
        SELECT *
        FROM links
        WHERE id=?`,
        [id],
      );
      await db.close();

      return new LinkDto(result);
    } catch (error) {
      throw new HttpException('can not make transaction', HttpStatus.INTERNAL_SERVER_ERROR);
    }
  }

  public async tagUnreachable(id: string, isReachable: boolean): Promise<iResult> {
    const db = await AsyncDatabase.open(process.env.DB_NAME);
    try {
      const result = await db.run(
        `
        UPDATE links
        SET is_reachable = ?,  date_update = ? 
        WHERE id=?`,
        [isReachable, dateConvert(), id],
      );
      await db.close();

      if (result.changes === 1) {
        return resultMaker(`One path tag as ${isReachable ? '' : 'un'}reachable`);
      }
      return resultMaker('No path tagged');
    } catch (error) {
      throw new HttpException('can not make transaction', HttpStatus.INTERNAL_SERVER_ERROR);
    }
  }

  public async addDuplicate(linkId: string, duplicateId: string): Promise<iResult> {
    const db = await AsyncDatabase.open(process.env.DB_NAME);
    try {
      const result = await db.run(
        `
        UPDATE links
        SET duplicate_id = ?, date_update = ?
        WHERE id = ?`,
        [duplicateId, dateConvert(), linkId],
      );
      await db.close();

      if (result.changes === 1) {
        return resultMaker(`Link ${linkId} tagged as duplicate of ${duplicateId}`);
      } else {
        return resultMaker('No path tagged as duplicate');
      }
    } catch (error) {
      throw new HttpException('can not make transaction', HttpStatus.INTERNAL_SERVER_ERROR);
    }
  }

  public async updateFilesNumber(
    id: number,
    { mediafiles, downloadedMediafiles, isDownloaded, progress }: UpdateLinkDto,
  ): Promise<iResult> {
    const db = await AsyncDatabase.open(process.env.DB_NAME);
    try {
      const result = await db.run(
        `
          UPDATE links 
          SET mediafiles = ?, downloaded_mediafiles = ?, is_downloaded = ?, progress = ?, date_update = ? 
          WHERE id = ?`,
        [mediafiles, downloadedMediafiles, isDownloaded, progress, dateConvert(), id],
      );
      await db.close();

      if (result.changes === 1) {
        return resultMaker('One path updated');
      }
      return resultMaker('No path updated');
    } catch (error) {
      throw new HttpException('can not make transaction', HttpStatus.INTERNAL_SERVER_ERROR);
    }
  }
}
